File: patterns/dotnet/AiPatterns/Domain/Models/UserProfile.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace AiPatterns.Domain.Models;

/// <summary>
/// User profile for MongoDB storage - demonstrates document/flexible schema data
/// </summary>
public class UserProfile
{
    [BsonId]
    [BsonRepresentation(BsonType.String)]
    public Guid Id { get; set; }

    [BsonElement("email")]
    public string Email { get; set; } = string.Empty;

    [BsonElement("firstName")]
    public string FirstName { get; set; } = string.Empty;

    [BsonElement("lastName")]
    public string LastName { get; set; } = string.Empty;

    [BsonElement("preferences")]
    public UserPreferences Preferences { get; set; } = new();

    [BsonElement("addresses")]
    public List<Address> Addresses { get; set; } = new();

    [BsonElement("paymentMethods")]
    public List<PaymentMethod> PaymentMethods { get; set; } = new();

    [BsonElement("tags")]
    public List<string> Tags { get; set; } = new();

    [BsonElement("metadata")]
    public Dictionary<string, object> Metadata { get; set; } = new();

    [BsonElement("createdAt")]
    public DateTime CreatedAt { get; set; }

    [BsonElement("updatedAt")]
    public DateTime UpdatedAt { get; set; }

    [BsonElement("lastLoginAt")]
    public DateTime? LastLoginAt { get; set; }

    public static UserProfile Create(string email, string firstName, string lastName)
    {
        return new UserProfile
        {
            Id = Guid.NewGuid(),
            Email = email,
            FirstName = firstName,
            LastName = lastName,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };
    }

    public void UpdatePreferences(UserPreferences newPreferences)
    {
        Preferences = newPreferences;
        UpdatedAt = DateTime.UtcNow;
    }

    public void AddAddress(Address address)
    {
        Addresses.Add(address);
        UpdatedAt = DateTime.UtcNow;
    }

    public void UpdateLastLogin()
    {
        LastLoginAt = DateTime.UtcNow;
        UpdatedAt = DateTime.UtcNow;
    }
}

public class UserPreferences
{
    [BsonElement("theme")]
    public string Theme { get; set; } = "system";

    [BsonElement("language")]
    public string Language { get; set; } = "en";

    [BsonElement("timezone")]
    public string Timezone { get; set; } = "UTC";

    [BsonElement("notifications")]
    public NotificationSettings Notifications { get; set; } = new();

    [BsonElement("privacy")]
    public PrivacySettings Privacy { get; set; } = new();

    [BsonElement("categories")]
    public List<string> FavoriteCategories { get; set; } = new();
}

public class NotificationSettings
{
    [BsonElement("email")]
    public bool EmailEnabled { get; set; } = true;

    [BsonElement("sms")]
    public bool SmsEnabled { get; set; } = false;

    [BsonElement("push")]
    public bool PushEnabled { get; set; } = true;

    [BsonElement("frequency")]
    public string Frequency { get; set; } = "daily";
}

public class PrivacySettings
{
    [BsonElement("profileVisible")]
    public bool ProfileVisible { get; set; } = true;

    [BsonElement("dataSharing")]
    public bool DataSharingEnabled { get; set; } = false;

    [BsonElement("analytics")]
    public bool AnalyticsEnabled { get; set; } = true;
}

public class Address
{
    [BsonElement("type")]
    public string Type { get; set; } = "shipping"; // shipping, billing

    [BsonElement("street")]
    public string Street { get; set; } = string.Empty;

    [BsonElement("city")]
    public string City { get; set; } = string.Empty;

    [BsonElement("state")]
    public string State { get; set; } = string.Empty;

    [BsonElement("zipCode")]
    public string ZipCode { get; set; } = string.Empty;

    [BsonElement("country")]
    public string Country { get; set; } = string.Empty;

    [BsonElement("isDefault")]
    public bool IsDefault { get; set; } = false;
}

public class PaymentMethod
{
    [BsonElement("type")]
    public string Type { get; set; } = string.Empty; // card, paypal, bank

    [BsonElement("provider")]
    public string Provider { get; set; } = string.Empty;

    [BsonElement("lastFour")]
    public string LastFour { get; set; } = string.Empty;

    [BsonElement("expiryMonth")]
    public int? ExpiryMonth { get; set; }

    [BsonElement("expiryYear")]
    public int? ExpiryYear { get; set; }

    [BsonElement("isDefault")]
    public bool IsDefault { get; set; } = false;

    [BsonElement("isActive")]
    public bool IsActive { get; set; } = true;
}

File: patterns/go/internal/domain/models/user_profile.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// UserProfile for MongoDB storage - demonstrates document/flexible schema data
type UserProfile struct {
	ID             uuid.UUID              `json:"id" bson:"_id"`
	Email          string                 `json:"email" bson:"email"`
	FirstName      string                 `json:"firstName" bson:"firstName"`
	LastName       string                 `json:"lastName" bson:"lastName"`
	Preferences    UserPreferences        `json:"preferences" bson:"preferences"`
	Addresses      []Address              `json:"addresses" bson:"addresses"`
	PaymentMethods []PaymentMethod        `json:"paymentMethods" bson:"paymentMethods"`
	Tags           []string               `json:"tags" bson:"tags"`
	Metadata       map[string]interface{} `json:"metadata" bson:"metadata"`
	CreatedAt      time.Time              `json:"createdAt" bson:"createdAt"`
	UpdatedAt      time.Time              `json:"updatedAt" bson:"updatedAt"`
	LastLoginAt    *time.Time             `json:"lastLoginAt,omitempty" bson:"lastLoginAt,omitempty"`
}

// UserPreferences holds user preference settings
type UserPreferences struct {
	Theme              string               `json:"theme" bson:"theme"`
	Language           string               `json:"language" bson:"language"`
	Timezone           string               `json:"timezone" bson:"timezone"`
	Notifications      NotificationSettings `json:"notifications" bson:"notifications"`
	Privacy            PrivacySettings      `json:"privacy" bson:"privacy"`
	FavoriteCategories []string             `json:"favoriteCategories" bson:"favoriteCategories"`
}

// NotificationSettings holds notification preferences
type NotificationSettings struct {
	Email       bool `json:"email" bson:"email"`
	Push        bool `json:"push" bson:"push"`
	SMS         bool `json:"sms" bson:"sms"`
	Marketing   bool `json:"marketing" bson:"marketing"`
	OrderStatus bool `json:"orderStatus" bson:"orderStatus"`
}

// PrivacySettings holds privacy preferences
type PrivacySettings struct {
	ProfilePublic     bool `json:"profilePublic" bson:"profilePublic"`
	ShowOnlineStatus  bool `json:"showOnlineStatus" bson:"showOnlineStatus"`
	AllowDataSharing  bool `json:"allowDataSharing" bson:"allowDataSharing"`
	AllowTracking     bool `json:"allowTracking" bson:"allowTracking"`
}

// Address represents a user address
type Address struct {
	ID         uuid.UUID `json:"id" bson:"id"`
	Label      string    `json:"label" bson:"label"` // home, work, etc.
	Street     string    `json:"street" bson:"street"`
	City       string    `json:"city" bson:"city"`
	State      string    `json:"state" bson:"state"`
	PostalCode string    `json:"postalCode" bson:"postalCode"`
	Country    string    `json:"country" bson:"country"`
	IsDefault  bool      `json:"isDefault" bson:"isDefault"`
}

// PaymentMethod represents a saved payment method
type PaymentMethod struct {
	ID           uuid.UUID `json:"id" bson:"id"`
	Type         string    `json:"type" bson:"type"` // card, paypal, etc.
	Last4        string    `json:"last4" bson:"last4"`
	ExpiryMonth  int       `json:"expiryMonth,omitempty" bson:"expiryMonth,omitempty"`
	ExpiryYear   int       `json:"expiryYear,omitempty" bson:"expiryYear,omitempty"`
	IsDefault    bool      `json:"isDefault" bson:"isDefault"`
	BillingEmail string    `json:"billingEmail,omitempty" bson:"billingEmail,omitempty"`
}

// NewUserProfile creates a new user profile with defaults
func NewUserProfile(email, firstName, lastName string) *UserProfile {
	now := time.Now().UTC()
	return &UserProfile{
		ID:        uuid.New(),
		Email:     email,
		FirstName: firstName,
		LastName:  lastName,
		Preferences: UserPreferences{
			Theme:    "system",
			Language: "en",
			Timezone: "UTC",
			Notifications: NotificationSettings{
				Email:       true,
				Push:        true,
				SMS:         false,
				Marketing:   false,
				OrderStatus: true,
			},
			Privacy: PrivacySettings{
				ProfilePublic:     false,
				ShowOnlineStatus:  true,
				AllowDataSharing:  false,
				AllowTracking:     false,
			},
			FavoriteCategories: []string{},
		},
		Addresses:      []Address{},
		PaymentMethods: []PaymentMethod{},
		Tags:           []string{},
		Metadata:       make(map[string]interface{}),
		CreatedAt:      now,
		UpdatedAt:      now,
	}
}

// UpdatePreferences updates the user's preferences
func (u *UserProfile) UpdatePreferences(prefs UserPreferences) {
	u.Preferences = prefs
	u.UpdatedAt = time.Now().UTC()
}

// AddAddress adds a new address to the user profile
func (u *UserProfile) AddAddress(addr Address) {
	if addr.ID == uuid.Nil {
		addr.ID = uuid.New()
	}
	u.Addresses = append(u.Addresses, addr)
	u.UpdatedAt = time.Now().UTC()
}

// UpdateLastLogin updates the last login timestamp
func (u *UserProfile) UpdateLastLogin() {
	now := time.Now().UTC()
	u.LastLoginAt = &now
	u.UpdatedAt = now
}

// GetFullName returns the user's full name
func (u *UserProfile) GetFullName() string {
	return u.FirstName + " " + u.LastName
}

// CreateUserRequest represents the request to create a user
type CreateUserRequest struct {
	Email     string `json:"email"`
	FirstName string `json:"firstName"`
	LastName  string `json:"lastName"`
}

// UpdatePreferencesRequest represents the request to update user preferences
type UpdatePreferencesRequest struct {
	Preferences UserPreferences `json:"preferences"`
}
